subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AntWorkAction.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.quality.internal;
+
+import groovy.lang.Closure;
+import org.apache.tools.ant.BuildListener;
+import org.apache.tools.ant.Project;
+import org.gradle.api.AntBuilder;
+import org.gradle.api.GradleException;
+import org.gradle.api.internal.project.ant.AntLoggingAdapter;
+import org.gradle.api.internal.project.ant.BasicAntBuilder;
+import org.gradle.api.internal.project.antbuilder.AntBuilderDelegate;
+import org.gradle.internal.jvm.Jvm;
+import org.gradle.util.internal.ClosureBackedAction;
+import org.gradle.workers.WorkAction;
+import org.gradle.workers.WorkParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Vector;
+
+public abstract class AntWorkAction<T extends WorkParameters> implements WorkAction<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AntWorkAction.class);
+
+    @Override
+    public void execute() {
+        LOGGER.info("Running {} with toolchain '{}'.", getActionName(), Jvm.current().getJavaHome().getAbsolutePath());
+        AntBuilder antBuilder = new BasicAntBuilder();
+        AntLoggingAdapter antLogger = new AntLoggingAdapter();
+        try {
+            configureAntBuilder(antBuilder, antLogger);
+            Object delegate = new AntBuilderDelegate(antBuilder, Thread.currentThread().getContextClassLoader());
+            ClosureBackedAction.execute(delegate, getAntClosure());
+        } finally {
+            disposeBuilder(antBuilder, antLogger);
+        }
+    }
+
+    protected abstract String getActionName();
+
+    protected abstract Closure<Object> getAntClosure();
```

Comment:
If I use `Action<AntBuilder>` directly I get `Class not found exception for: com.puppycrawl.tools.checkstyle.CheckStyleTask` even though the classpath set for worker contains the `Checkstyle`. So it seems that for some reason system classloader doesn't contain this class, but the current thread classloader has it.

So it might be an issue with the worker API.  I didn't research it any further.

Reply:
Is this like what you tried? This passes for me:
https://github.com/gradle/gradle/commit/9f230aeb269e480efd9b369f4424d66b3ac30d99

Reply:
It seems that it works. I probably tried to use AntBuilder directly without the delegate. Thanks for looking into it.